renku/core/dataset/dataset.py

```diff
@@ -1168,3 +1170,90 @@ def should_include(filepath: Path) -> bool:
         raise errors.ParameterError(f"These datasets don't exist: {unused_names_str}")
 
     return sorted(records, key=lambda r: r.date_added)
+
+
+def pull_external_data(name: str, location: Optional[Path] = None) -> None:
+    """Pull/copy data for an external storage to a dataset's data directory or a specified location.
+
+    Args:
+        name(str): Name of the dataset
+        location(Optional[Path]): A directory to copy data to (Default value = None).
+    """
+    client = get_client()
+    datasets_provenance = DatasetsProvenance()
+
+    dataset = datasets_provenance.get_by_name(name=name, strict=True)
+    assert dataset  # NOTE: To make mypy happy
+
+    if not dataset.storage:
+        communication.warn(f"Dataset '{name}' doesn't have a storage backend")
+        return
+
+    create_symlinks = True
+    destination: Union[Path, str]
+
+    if location:
+        destination = get_absolute_path(location)
+    else:
+        stored_location = read_dataset_data_location(dataset=dataset)
+        if stored_location:
+            destination = stored_location
+        else:
+            destination = client.path
+            create_symlinks = False
+
+    provider = ProviderFactory.get_pull_provider(uri=dataset.storage)
+
+    credentials = S3Credentials(provider)
+    prompt_for_credentials(credentials)
+
+    storage = get_storage(provider=provider, credentials=credentials)
+    updated_files = []
+
+    for file in dataset.files:
+        path = Path(destination) / file.entity.path
+        path.parent.mkdir(parents=True, exist_ok=True)
+        # NOTE: Don't check if destination exists. ``IStorage.copy`` won't copy a file if it exists and is not modified.
+
+        if not file.source:
+            raise errors.DatasetImportError(f"Dataset file doesn't have a URI: {file.entity.path}")
+
+        with communication.busy(f"Copying {file.entity.path} ..."):
+            storage.copy(file.source, path)
+
+            if file.based_on and not file.based_on.checksum:
+                md5_hash = hash_file(path, hash_type="md5") or ""
+                file.based_on = RemoteEntity(checksum=md5_hash, url=file.based_on.url, path=file.based_on.path)
+
+            new_file = DynamicProxy(file)
+            new_file.dataset = dataset
+            updated_files.append(new_file)
+
+            if create_symlinks:
+                symlink_path = client.path / file.entity.path
+                symlink_path.parent.mkdir(parents=True, exist_ok=True)
+                create_symlink(path=path, symlink_path=symlink_path, overwrite=True)
+
+    # NOTE: Store location in metadata in case where we want to mount the external storage in the same location
+    store_dataset_data_location(dataset=dataset, location=location)
+
+    if updated_files:
+        _update_datasets_files_metadata(client, updated_files=updated_files, deleted_files=[], delete=False)
```

Comment:
Should we only update metadata if a file was actually changed during the pull?

Reply:
When we add data from S3, it's not always possible to get the checksum from S3. So, we leave an empty checksum in the metadata and then we calculate the checksum here once we've downloaded the data. This is for such cases.